src/block/block_compact.c

```diff
@@ -543,12 +544,16 @@ __compact_page_skip(
 
     /*
      * We must have reviewed at least some interesting number of pages for any estimates below to be
-     * worthwhile.
+     * worthwhile. If compaction has rewritten more pages than expected, estimate the work again.
      */
+    if (block->compact_estimated &&
+      block->compact_pages_rewritten > block->compact_pages_rewritten_expected)
+        block->compact_estimated = false;
     if (!block->compact_estimated && block->compact_pages_reviewed >= WT_THOUSAND) {
+        compact_pages_rewritten_expected_prev = block->compact_pages_rewritten_expected;
         __block_compact_estimate_remaining_work(session, block);
         /* If no potential work has been found, exit compaction. */
-        if (block->compact_pages_rewritten_expected == 0)
+        if (compact_pages_rewritten_expected_prev == block->compact_pages_rewritten_expected)
```

Comment:
I would keep this condition as it was. If the estimation returns the same non-zero result as last time, it just means that the re-estimation wasn't particularly useful; it is worth to finish the compaction itself, as it is already saving us up some space.

Reply:
That's very true, thank you for spotting this. Fixed in 8ddd3d396d080064a6f425a1e6e5abf6f4b0396a.